[Synthetics] Decouple synthetics settings from uptime

shahzad31 opened the pull request:
## Summary

Fixes https://github.com/elastic/kibana/issues/183916 

Decouple synthetics settings from uptime , since uptime isn't available in serverless !!

### Test cases

- [ ] Make sure synthetics settings continue to work as expected in serverless/stateful
- [ ] Make sure uptime settings continues to work as expected in serverless/stateful
- [ ] if User have uptime settings saved, those should be copied over to synthetics first time

shahzad31 committed:
wip

shahzad31 committed:
Merge branch 'main' of https://github.com/elastic/kibana into synthetics-settings

shahzad31 committed:
fix settings

elasticmachine commented:
Pinging @elastic/obs-ux-infra_services-team (Team:obs-ux-infra_services)

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `run` `docs-build` : Re-trigger the docs validation. (use unformatted text in the comment!)
</p>
</details>

shahzad31 committed:
handle error

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/check_mappings_update --fix'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/jest_integration -u src/core/server/integration_tests/ci_checks'

justinkambic reviewed (changes requested):
I didn't seem to find any issues with the code when I read through it, but from a functionality standpoint I discovered an issue (maybe).

I created a server log connector and added it to Uptime. I verified that this was mirrored in Synthetics per the PR description, and when I deleted it from Synthetics and tried to persist the change I got a 500.

<img width="1660" alt="image" src="https://github.com/elastic/kibana/assets/18429259/0b9de11b-1600-42ed-8f5e-f6a138f7768d">

The only message logged by Kibana was:

```bash
[2024-05-22T16:24:19.533-04:00][ERROR][http] 500 Server Error
```

shahzad31 committed:
Merge branch 'main' of https://github.com/elastic/kibana into synthetics-settings

shahzad31 committed:
update tests

jloleysens commented on x-pack/plugins/observability_solution/synthetics/server/saved_objects/synthetics_settings.ts:
```diff
@@ -0,0 +1,47 @@
+  mappings: {
+    dynamic: false,
+    properties: {
+      /* Leaving these commented to make it clear that these fields exist, even though we don't want them indexed.
+         When adding new fields please add them here. If they need to be searchable put them in the uncommented
+         part of properties.
+      certAgeThreshold: {
+        type: 'long',
+      },
+      certExpirationThreshold: {
+        type: 'long',
+      },
+      defaultConnectors: {
+        type: 'keyword',
+      },
+      */
+    },
+  },
```
Instead of maintaining a comment you can consider adding model version `1` with `changes: []` (since there are no changes, this is first version) and specifying a schema that describes the contents of the document:

https://github.com/elastic/kibana/blob/6150a221c26369a67031fbf5d9147b231372f31a/src/plugins/data/server/saved_objects/query.ts#L38

shahzad31 committed:
Merge branch 'main' of https://github.com/elastic/kibana into synthetics-settings

shahzad31 committed:
fix types

kc13greiner commented on x-pack/plugins/observability_solution/synthetics/server/routes/settings/dynamic_settings.ts:
```diff
@@ -0,0 +1,82 @@
+import { DynamicSettingsAttributes } from '../../runtime_types/settings';
+import { SYNTHETICS_API_URLS } from '../../../common/constants';
+
+export const createGetDynamicSettingsRoute: SyntheticsRestApiRouteFactory<
```
Question: Should any of these new routes be protected with access tags?

shahzad31 replied:
@kc13greiner app has logic to by default add write tags to all non-GET routes

kc13greiner reviewed (commented):
Looking good overall - just one question!

shahzad31 committed:
update api tests

shahzad31 committed:
update types

shahzad31 committed:
Merge branch 'main' of https://github.com/elastic/kibana into synthetics-settings

shahzad31 commented:
> I didn't seem to find any issues with the code when I read through it, but from a functionality standpoint I discovered an issue (maybe).
> 
> I created a server log connector and added it to Uptime. I verified that this was mirrored in Synthetics per the PR description, and when I deleted it from Synthetics and tried to persist the change I got a 500.
> 
> <img alt="image" width="1660" src="https://private-user-images.githubusercontent.com/18429259/332949735-0b9de11b-1600-42ed-8f5e-f6a138f7768d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY1NjQ3NjYsIm5iZiI6MTcxNjU2NDQ2NiwicGF0aCI6Ii8xODQyOTI1OS8zMzI5NDk3MzUtMGI5ZGUxMWItMTYwMC00MmVkLThmNWUtZjZhMTM4Zjc3NjhkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI0VDE1Mjc0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI1NWE4NTVkZTcwMGQ0MjQ2ZmM3YTAzYjRkOGI4YzgzYzgyZmY0MzcyNzRjNzYyYTE4ZGQ1MzY2NjZlZmI0MjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WXeIbDQGkwRC0nuT588NI-r0UD-YQdRzxtusDVrj_rs">
> The only message logged by Kibana was:
> 
> ```shell
> [2024-05-22T16:24:19.533-04:00][ERROR][http] 500 Server Error
> ```

This should be tried on fresh ES cluster

dominiqueclarke commented on x-pack/plugins/observability_solution/synthetics/server/saved_objects/saved_objects.ts:
```diff
@@ -52,33 +53,49 @@ export const registerUptimeSavedObjects = (
-        }
-        return DYNAMIC_SETTINGS_DEFAULT_ATTRIBUTES;
+        // If the object doesn't exist, check to see if uptime settings exist
+        return getUptimeDynamicSettings(client);
```
I would remove this. It creates an confusing tie between Uptime and Synthetics in my opinion.

For example, the first time I used settings things were mirrored between Uptime and Synthetics. For the next uses, I expected that mirroring to continue, but it did not. 

The mirroring was due to this login where I initially had only an uptime setting SO and not a Synthetics one.

I would love to avoid falling back to Uptime, but I suppose we need this in place to account for the connection we currently have between the two? Is there another option? Some way to migrate the old Uptime settings into a new Synthetics SO once and then remove the need for this fallback?

shahzad31 replied:
unfortunately there is no other way.

shahzad31 committed:
Merge branch 'main' of https://github.com/elastic/kibana into synthetics-settings

shahzad31 commented on x-pack/plugins/observability_solution/synthetics/server/lib.test.ts:
```diff
@@ -42,7 +42,7 @@ describe('UptimeEsClient', () => {
       expect(esClient.search).toHaveBeenCalledWith(
         {
-          index: 'heartbeat-*',
+          index: 'synthetics-*',
```
i have removed it already. this is just a test.

shahzad31 commented on x-pack/plugins/observability_solution/synthetics/server/legacy_uptime/lib/requests/get_network_events.test.ts:
```diff
@@ -213,7 +213,7 @@ describe('getNetworkEvents', () => {
               "track_total_hits": true,
             },
-            "index": "heartbeat-*",
+            "index": "synthetics-*",
```
well probably needs some refactoring.

shahzad31 commented on x-pack/plugins/observability_solution/synthetics/server/saved_objects/saved_objects.ts:
```diff
@@ -52,33 +53,49 @@ export const registerUptimeSavedObjects = (
+  ) => {
+    const settingsObject = await client.create<DynamicSettingsAttributes>(
+      syntheticsSettingsObjectType,
+      settings,
```
but we are using route schema validation, so i am not sure how it will add other keys?

shahzad31 committed:
fix types

shahzad31 committed:
Merge branch 'main' of https://github.com/elastic/kibana into synthetics-settings

shahzad31 committed:
fix types

shahzad31 committed:
update snaps

dominiqueclarke reviewed (approved):
LGTM
